embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceApiService.kt

```diff
@@ -170,38 +172,45 @@ internal class EmbraceApiService(
         return postOnExecutor(bytes, request, onComplete)
     }
 
+    /**
+     * Posts an API call to the API.
+     * If the device is offline or the endpoint is rate limited, the API call is saved to be sent later.
+     * If the API call fails, it is retried and if it succeeds, it is removed from cache.
+     */
     private fun postOnExecutor(
         payload: ByteArray,
         request: ApiRequest,
         onComplete: (() -> Any)?
     ): Future<*> {
         return executorService.submit(
             NetworkRequestRunnable(request) {
-                try {
-                    if (lastNetworkStatus != NetworkStatus.NOT_REACHABLE) {
-                        executePost(request, payload)
+                val pendingApiCall = pendingApiCallsSender.savePendingApiCall(request, payload)
+
+                if (lastNetworkStatus != NetworkStatus.NOT_REACHABLE &&
+                    !rateLimitHandler.isRateLimited(request.url.endpoint())
+                ) {
+                    // Execute the request if the device is online and the endpoint is not rate limited.
+                    val response: ApiResponse = executePost(request, payload)
+
+                    if (response is ApiResponse.Success) {
+                        // Remove the API call from cache if it was sent successfully.
+                        pendingApiCallsSender.removePendingApiCall(pendingApiCall)
+                        onComplete?.invoke()
                     } else {
-                        pendingApiCallsSender.scheduleApiCall(request, payload)
-                        logger.logWarning("No connection available. Request was queued to retry later.")
+                        // Schedule the API call to be sent later if it should be retried.
+                        if (pendingApiCallsSender.shouldRetry(response)) {
+                            pendingApiCallsSender.scheduleApiCall(response)
+                        }
+                        onComplete?.invoke()
```

Comment:
The reason of removing the try-catch-finally is that this piece of code doesn't throw exceptions any more, try-catch is used in ApiClient and when an exception is cought we return to the ApiService a ApiResponse.Incomplete object. 
With other methods called in this function is the same, they don't throw anything, they try-catch inside.

Reply:
So `onComplete()` is invoked before regardless of what the status was because it was in the `finally` block. If you kept the same pattern (but got rid of the catch block), you wouldn't have to call that in two places.

Ah, and good to know `executePost()` already contains try-catch logic. Good to know.

Reply:
The idea behind that was to avoid so many nested blocks, but you are right that calling the `onComplete()` in just one place is much better. 
I added the `try-finally` block but I created a separate function to simplify and make more readable that part of the code.